### PR TITLE
Packed Secret Sharing Scheme

### PR DESCRIPTION
**Pull request by sdiehl**

### CODE REVIEW

File: src/Shamir.hs [hunk 1]
```diff
@@ -1,6 +1,7 @@
 {-# LANGUAGE DeriveAnyClass #-}
 {-# LANGUAGE DeriveGeneric #-}
 {-# LANGUAGE ParallelListComp #-}
+-- Shamir secret sharing using Lagrange interpolation
```
Comment (sdiehl) [inline src/Shamir.hs]: Use a module docstring.
https://www.haskell.org/haddock/doc/html/ch03s03.html

File: src/Shamir/FFT.hs [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+  where
+    largeValues = 0 : shares
+    largeCoeffs = inverseDft3 omega3 largeValues
+    smallCoeffs = take (length shares) largeCoeffs
+    smallValues = fft2 omega2 smallCoeffs
+    secrets = take k (drop 1 smallValues)
+
+-- -- | Create shares from a secret
+-- -- See https://mortendahl.github.io/2017/06/24/secret-sharing-part2/
+-- shareSecret :: (MonadRandom m, PrimeField f) => f -> Int -> Int -> m [f]
```
Comment (sdiehl) [inline src/Shamir/FFT.hs]: Remove dead code?

File: ChangeLog.md [hunk 3]
```diff
@@ -1,5 +1,9 @@
 # Changelog for shamir
 
+## 0.2
```
Comment (sdiehl) [inline ChangeLog.md]: Just 0.1

File: src/Shamir/Packed.hs [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+-------------------------------------
+-- Newton Polynomial Interpolation --
+-------------------------------------
+
+data NewtonPolynomial f = NewtonPolynomial
+  { npPoints :: [f]
+  , npCoeffs :: [f]
+  } deriving (Show, Eq)
+
+newtonInterpolation :: PrimeField f => [f] -> [f] -> NewtonPolynomial f
```
Comment (sdiehl) [inline src/Shamir/Packed.hs]: Nice 🙂

Comment (sdiehl) [reply src/Shamir/Packed.hs]: List lookups using ! are quite slow. This is the completely unoptimized version?

Comment (Acentelles) [reply src/Shamir/Packed.hs]: Yes, we can do better. Let's use a map instead and add some benchmarks before merging

**Pull request merged**